### PR TITLE
Re-host the dart web UI on v3 of the server protocol

### PR DESCRIPTION
**Pull request by devoncarew**
- re-host the dart web UI on v3 of the server protocol

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by devoncarew** [comment]
@johnpryan - no hurry on this; I'll likely have some questions about the workshops UI and testing mode code before landing this.

File: pkgs/dart_pad/lib/elements/analysis_results_controller.dart
Comment (johnpryan) [inline pkgs/dart_pad/lib/elements/analysis_results_controller.dart:121]: Why is this commented out?

File: pkgs/dart_pad/lib/elements/analysis_results_controller.dart
Comment (johnpryan) [inline pkgs/dart_pad/lib/elements/analysis_results_controller.dart:188]: Why is this commented out?

File: pkgs/dart_pad/lib/embed.dart
Comment (johnpryan) [inline pkgs/dart_pad/lib/embed.dart:317]: This is required for codelabs like the [layout basics](https://docs.flutter.dev/codelabs/layout-basics) codelab. Is there a reason this is commented out?

File: pkgs/dart_pad/lib/embed.dart
Comment (johnpryan) [inline pkgs/dart_pad/lib/embed.dart:926]: Same comment as above

**Review by johnpryan — CHANGES_REQUESTED** [review]
It would be good to keep the diagnostic messages and the test code features working for now. Is that something you can do in this PR?

File: pkgs/dart_pad/lib/embed.dart
Comment (devoncarew) [reply pkgs/dart_pad/lib/embed.dart:317]: I added this back in. When updating this frontend, I'd assumed we didn't actually have more then one file. Re-reading this, it looks like we have two tabs (src and tests), which are analyzed as one file, but then any analysis issues reassigned back into two files after.

I added this back.

File: pkgs/dart_pad/lib/embed.dart
Comment (devoncarew) [reply pkgs/dart_pad/lib/embed.dart:926]: Added back in.

File: pkgs/dart_pad/lib/elements/analysis_results_controller.dart
Comment (devoncarew) [reply pkgs/dart_pad/lib/elements/analysis_results_controller.dart:188]: We no longer pass back an analysis issues context messages (here, called diagnostics). These are optional metadata about an issue. The user has enough to go on with the analysis issue type, message, and location. (so, I removed this UI as we no longer have the model information that is was using)

File: pkgs/dart_pad/lib/elements/analysis_results_controller.dart
Comment (devoncarew) [reply pkgs/dart_pad/lib/elements/analysis_results_controller.dart:121]: Same as the other comment - I removed the `contextMessages` field from the info about an analysis issue.

**Review by devoncarew — COMMENTED** [review]
Updated! What's a good way to test the codelabs locally?

**Comment by johnpryan** [comment]
You can test the codelabs and workshops by loading the right URL. For example:

To test a workshop you can click on one of the [links in the README](https://github.com/flutter/codelabs/tree/main/dartpad_codelabs) and copy-paste the URL after the `/`. For example: `localhost:8080/workshops.html?webserver=https://dartpad-workshops-io2021.web.app/inherited_widget`

To test a codelab you can right-click and grab the `src` attribute on the embedded DartPad iframe from the codelab such as the [Basic Flutter layout concepts](https://docs.flutter.dev/codelabs/layout-basics) codelab, and do the same thing. For example `http://localhost:8080/embed-flutter.html?theme=dark&run=false&split=60&ga_id=false`

**Comment by johnpryan** [comment]
You can also verify the embeds are working using the example at `http://localhost:8080/example/embed.html`

**Comment by devoncarew** [comment]
> It would be good to keep the diagnostic messages and the test code features working for now. Is that something you can do in this PR?

I can definitely add back in the test code features - I'd assumed from reading through the code that it wasn't used, but was mistaken.

Do you think the diagnostic messages functionality is important? It's not currently exposed in the new UI, and I wasn't really planning to do it. I think it's useful for more sophisticated diagnostics in an IDE, but I don't think critical here.

**Comment by johnpryan** [comment]
> Do you think the diagnostic messages functionality is important?

IIRC this was to enable corrections and documentation links: https://github.com/dart-lang/dart-pad/issues/1676

**Comment by devoncarew** [comment]
I restore the `contextMessages` info and related UI; this PR still needs some manual testing through the codelabs and embeds.

**Comment by johnpryan** [comment]
> this PR still needs some manual testing through the codelabs and embeds.

I'm seeing an issue with all three versions where the analysis message is always the same:

Screenshot 2023-11-15 at 8 53 13 AM

**Comment by devoncarew** [comment]
> I'm seeing an issue with all three versions where the analysis message is always the same:

The protocol changed a bit in this PR (due to the inclusion context messages and changing how we stored location info on analysis issues); you'll need to run this against a local backend to see results.

That's a good reminder though; we may want to land the dart-services + sketchpad portion of this PR first, and follow up ~an hour later with the part moving the current UI over to the new backend.

**Comment by devoncarew** [comment]
Codelabs and the tests UI both seem to be working (responding to the server, showing analysis issues, supporting run, ...).

**Review by johnpryan — APPROVED** [review]

**Comment by johnpryan** [comment]
We might want to update CONTRIBUTING.md and other readme files to remove any references to `protoc`

**Comment by devoncarew** [comment]
> We might want to update CONTRIBUTING.md and other readme files to remove any references to `protoc`

👍 ; I'd like to do that in a follow-up (this PR has accumulated a bunch of changes).

I'll hold this PR until we determine the build issue with the master channel (we'll need that server to support the new protocol).

**Comment by johnpryan** [comment]
> I'll hold this PR until we determine the build issue with the master channel (we'll need that server to support the new protocol).

The checks on the `main` channel of the repo are passing, it's only the Cloud Run deploy action that's failing so this should be okay to merge.

**Comment by devoncarew** [comment]
> The checks on the `main` branch of the repo are passing, it's only the Cloud Run deploy action for the Flutter main channel that's failing so this should be okay to merge.

I think the checks were unintentionally all run against `stable` (which passed happily). The regression on `main` has now made its way to `beta`.

**Pull request merged**
